File: retaRust/python_arch_reference/reta_architecture/output_syntax.py
from __future__ import annotations

"""Output syntax morphisms for Reta table renderers.

Stage 24 extracts the concrete renderer syntax classes from the old
``libs.lib4tables`` module. These classes are now owned by the architecture
layer; ``libs.lib4tables`` remains a small compatibility facade for legacy
imports.
"""

from collections import OrderedDict
from dataclasses import dataclass

from .number_theory import primCreativity
from .split_i18n import build_split_i18n_proxy

_i18n_proxy = build_split_i18n_proxy()
i18n = _i18n_proxy.lib4tables

class NichtsSyntax:
    mode_name = "nichts"
    force_one_table = False
    force_zero_width = False
    marks_html_or_bbcode = False
    def coloredBeginCol(self, num: int, rest: bool = False):
        return ""

    def generateCell(
        self, num: int, dataDict: dict, content=None, zeile=None, tables=None
    ) -> str:
        return ""

    beginTable = ""
    endTable = ""
    beginCell = ""
    endCell = ""
    beginZeile = ""
    endZeile = ""


class OutputSyntax:
    mode_name = "shell"
    force_one_table = False
    force_zero_width = False
    marks_html_or_bbcode = False
    def coloredBeginCol(self, num: int, rest: bool = False):
        return self.beginZeile

    def generateCell(
        self, num: int, dataDict: dict, content=None, zeile=None, tables=None
    ) -> str:
        return self.beginCell

    beginTable = ""
    endTable = ""
    beginCell = ""
    endCell = ""
    beginZeile = ""
    endZeile = ""


class csvSyntax(OutputSyntax):
    mode_name = "csv"
    force_one_table = True
    force_zero_width = True
    beginTable = ""
    endTable = ""
    beginCell = ""
    endCell = ""
    beginZeile = ""
    endZeile = ""


class emacsSyntax(OutputSyntax):
    mode_name = "emacs"
    force_one_table = True
    force_zero_width = True
    # print(OutputSyntax)
    # def generateCell(
    #    self, spalte: int, SpaltenParameter: dict, content=None, zeile=None, tables=None
    # ) -> str:
    #    return "|"

    beginTable = ""
    endTable = ""
    beginCell = "|"
    endCell = ""
    beginZeile = ""
    endZeile = "|"


class markdownSyntax(OutputSyntax):
    mode_name = "markdown"
    force_one_table = True
    force_zero_width = True
    # print(OutputSyntax)
    # def generateCell(
    #    self, spalte: int, SpaltenParameter: dict, content=None, zeile=None, tables=None
    # ) -> str:
    #    return "|"

    beginTable = ""
    endTable = ""
    beginCell = "|"
    endCell = ""
    beginZeile = ""
    endZeile = "|"


class bbCodeSyntax(OutputSyntax):
    mode_name = "bbcode"
    marks_html_or_bbcode = True
    def coloredBeginCol(self, num: int, rest: bool = False):
        num = int(num) if str(num).isdecimal() else 0
        numberType = primCreativity(num)

        if rest:
            # wenn der Fallm eintritt dass es leerer Text ist der frei ist
            return "[tr]"
            if num == 0:
                return "[tr]"
            elif num % 2 == 0:
                return "[tr]"
            else:
                return "[tr]"
        elif numberType == 1:
            if num % 2 == 0:
                return '[tr="background-color:#66ff66;color:#000000;"]'
            else:
                return '[tr="background-color:#009900;color:#ffffff;"]'
        elif numberType == 2 or num == 1:
            if num % 2 == 0:
                return '[tr="background-color:#ffff66;color:#000099;"]'
            else:
                return '[tr="background-color:#555500;color:#aaaaff;"]'
        elif numberType == 3:
            if num % 2 == 0:
                return '[tr="background-color:#9999ff;color:#202000;"]'
            else:
                return '[tr="background-color:#000099;color:#ffff66;"]'
        elif num == 0:
            return '[tr="background-color:#ff2222;color:#002222;"]'

    def generateCell(
        self, spalte: int, SpaltenParameter: dict, content=None, zeile=None, tables=None
    ) -> str:
        spalte = int(spalte)
        spalte += 2
        return "".join(
            (
                "[td",
                (
                    '="background-color:#000000;color:#ffffff"'
                    if content is not None and int(content) % 2 == 0
                    else '="background-color:#ffffff;color:#000000"'
                )
                if spalte == 0
                else '=""',
                "]",
            )
        )

    beginTable = "[table]"
    endTable = "[/table]"
    beginCell = "[td]"
    endCell = "[/td]"
    beginZeile = "[tr]"
    endZeile = "[/tr]"


class htmlSyntax(OutputSyntax):
    mode_name = "html"
    marks_html_or_bbcode = True
    def __init__(self):
        self.zeile = 0

    def coloredBeginCol(self, num: int, rest: bool = False) -> str:
        num = int(num) if str(num).isdecimal() else 0
        numberType = primCreativity(num)
        self.zeile = num
        if rest:
            # wenn der Fallm eintritt dass es leerer Text ist der frei ist
            return "<tr>\n"
            if num == 0:
                return "<tr>\n"
            elif num % 2 == 0:
                return "<tr>\n"
            else:
                return "<tr>\n"
        elif numberType == 1:
            if num % 2 == 0:
                return '<tr style="background-color:#66ff66;color:#000000;">\n'
            else:
                return '<tr style="background-color:#009900;color:#ffffff;">\n'
        elif numberType == 2 or num == 1:
            if num % 2 == 0:
                return '<tr style="background-color:#ffff66;color:#000099;">\n'
            else:
                return '<tr style="background-color:#555500;color:#aaaaff;">\n'
        elif numberType == 3:
            if num % 2 == 0:
                return '<tr style="background-color:#9999ff;color:#202000;">\n'
            else:
                return '<tr style="background-color:#000099;color:#ffff66;">\n'
        elif num == 0:
            return '<tr style="background-color:#ff2222;color:#002222;">\n'

    def generateCell(
        self, spalte: int, SpaltenParameter: dict, content=None, zeile=None, tables=None
    ) -> str:
        if zeile == "":
            zeile = 0
        spalte = int(spalte)
        if spalte == -2:
            tupleOfListsOfCouples = (
                [
                    (i18n.zaehlung["Zählung"], ""),
                ],
            )
        elif spalte == -1:
            tupleOfListsOfCouples = (
                [
                    (i18n.nummerier["Nummerierung"], ""),
                ],
            )
        else:
            try:
                # print(SpaltenParameter[spalte])
                tupleOfListsOfCouples = SpaltenParameter[spalte]
            except:
                if str(spalte).isdecimal():
                    raise ValueError
                tupleOfListsOfCouples = (("?", "?"),)
        # damit pypy3 == python3
        things1: OrderedDict[int, list] = OrderedDict()

        listOfListsOfCouples: list = list(tupleOfListsOfCouples)
        # listOfListsOfCouples.sort(
        #    key=lambda x: sorted([(s.upper(), v) for s, v in x])
        # )  # damit pypy3 == python3
        # print(listOfListsOfCouples)
        for c, couples in enumerate(listOfListsOfCouples):
            for paraNum in (0, 1):
                if len(couples[0]) > paraNum:
                    if len(couples[0]) > paraNum:
                        # i = 0
                        para1o2name = couples[0][paraNum]
                        # while (
                        #    len(couples) > i + 1
                        #    and len(couples[i + 1]) > 0
                        #    and para1o2name.strip() == ""
                        # ):
                        #    i += 1
                        #    para1o2name = couples[i][paraNum]
                        if len(para1o2name.strip()) != 0 or True:
                            if paraNum == 1:
                                para1o2name = "".join(["p3_", str(c), "_", para1o2name])
                            try:
                                things1[paraNum] += [para1o2name]
                                # things1[paraNum].sort(key=lambda x: x.upper())

                            except KeyError:
                                things1[paraNum]: list = [
                                    para1o2name,
                                ]

        things: OrderedDict = OrderedDict()  # damit pypy3 == python3

        for key, values in things1.items():
            for i, el in enumerate(values):
                if el != i18n.alles["alles"]:
                    try:
                        things[key] += (
                            "✗" if key == 0 else "",
                            el,
                            ",",
                        )
                    except KeyError:
                        things[key] = (
                            "✗" if key == 0 else "",
                            el,
                            ",",
                        )
            things[key] = "".join(things[key])

        spalte += 2
        if len(things) < 2:
            return ""
        else:
            p4: str
            try:
                p4a = tables.generatedSpaltenParameter_Tags[spalte - 2]
                # print(str(p4a))
                p4b: list = []
                for a in p4a:
                    p4b += [str(a.value)]
                # p4b.sort(key=lambda x: x.upper())
                p4 = ",".join(p4b)
            except KeyError:
                p4 = ""
            except:
                p4 = ""
            return "".join(
                ("<td",)
                + (
                    (
                        ' class="',
                        "z_",
                        str(int(zeile)),
                        " r_",
                        str(spalte),
                        " p1_",
                        things[0],
                        ",",
                        " p2_",
                        (things[1] if len(things) > 1 else ""),
                        " p4_",
                        p4,
                        '"',
                    )
                    if zeile == 0
                    else ()
                )
                + (
                    (
                        ' style="background-color:#000000;color:#ffffff;"'
                        if content is not None and int(content) % 2 == 0
                        else ' style="background-color:#ffffff;color:#000000;"'
                        if spalte == 0
                        else "",
                    )
                    if spalte in (0, 1)
                    # else (' style="display:none"',)
                    # if zeile == 0
                    else (
                        ' class="tdSymbole" style="background-image: url('
                        ');background-size: cover;background-repeat: no-repeat;background-position: right; "',
                    )
                    if "Symbole" in things1[0]
                    else ()
                )
                + (">\n",)
            )

    beginTable = '<table border=0 id="bigtable">'
    endTable = "</table>\n"
    beginCell = "<td>\n"
    endCell = "\n</td>\n"
    # beginZeile = "<tr>"
    beginZeile = ""
    endZeile = "</tr>\n"


OUTPUT_SYNTAX_CLASSES = {
    "shell": OutputSyntax,
    "nichts": NichtsSyntax,
    "csv": csvSyntax,
    "bbcode": bbCodeSyntax,
    "html": htmlSyntax,
    "emacs": emacsSyntax,
    "markdown": markdownSyntax,
}


@dataclass(frozen=True)
class OutputSyntaxBundle:
    classes: dict[str, type]

    def snapshot(self) -> dict:
        return {
            "class": "OutputSyntaxBundle",
            "modes": {
                name: {
                    "class": cls.__name__,
                    "force_one_table": bool(getattr(cls, "force_one_table", False)),
                    "force_zero_width": bool(getattr(cls, "force_zero_width", False)),
                    "marks_html_or_bbcode": bool(getattr(cls, "marks_html_or_bbcode", False)),
                }
                for name, cls in sorted(self.classes.items())
            },
            "legacy_owner": "libs.lib4tables",
            "architecture_owner": "reta_architecture.output_syntax",
        }

    def class_for(self, mode: str) -> type:
        return self.classes[mode]


def bootstrap_output_syntax() -> OutputSyntaxBundle:
    return OutputSyntaxBundle(classes=dict(OUTPUT_SYNTAX_CLASSES))


def output_syntax_snapshot() -> dict:
    return bootstrap_output_syntax().snapshot()


__all__ = [
    "NichtsSyntax",
    "OutputSyntax",
    "csvSyntax",
    "emacsSyntax",
    "markdownSyntax",
    "bbCodeSyntax",
    "htmlSyntax",
    "OUTPUT_SYNTAX_CLASSES",
    "OutputSyntaxBundle",
    "bootstrap_output_syntax",
    "output_syntax_snapshot",
]

File: retaRust/crates/reta_architecture/src/output_syntax.rs
//! Output syntax morphisms transcompiled from
//! `python_arch_reference/reta_architecture/output_syntax.py`.
//!
//! This module owns the renderer mode vocabulary and the mode-specific table,
//! row and cell wrappers.  The heavy historical renderer can keep producing
//! bytes for now; this layer gives Rust code the same syntax choices in typed
//! form.

use serde::{Deserialize, Serialize};

use crate::number_theory::prime_creativity;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize)]
pub enum OutputMode {
    Shell,
    Nichts,
    Csv,
    Bbcode,
    Html,
    Emacs,
    Markdown,
}

impl OutputMode {
    pub const fn canonical_name(self) -> &'static str {
        match self {
            Self::Shell => "shell",
            Self::Nichts => "nichts",
            Self::Csv => "csv",
            Self::Bbcode => "bbcode",
            Self::Html => "html",
            Self::Emacs => "emacs",
            Self::Markdown => "markdown",
        }
    }

    pub const fn syntax_class_name(self) -> &'static str {
        match self {
            Self::Shell => "OutputSyntax",
            Self::Nichts => "NichtsSyntax",
            Self::Csv => "csvSyntax",
            Self::Bbcode => "bbCodeSyntax",
            Self::Html => "htmlSyntax",
            Self::Emacs => "emacsSyntax",
            Self::Markdown => "markdownSyntax",
        }
    }

    pub const fn force_one_table(self) -> bool {
        matches!(self, Self::Csv | Self::Emacs | Self::Markdown)
    }

    pub const fn force_zero_width(self) -> bool {
        matches!(self, Self::Csv | Self::Emacs | Self::Markdown)
    }

    pub const fn marks_html_or_bbcode(self) -> bool {
        matches!(self, Self::Bbcode | Self::Html)
    }

    pub fn from_name(value: &str) -> Option<Self> {
        match value.trim().to_ascii_lowercase().as_str() {
            "shell" => Some(Self::Shell),
            "nichts" | "nothing" | "none" => Some(Self::Nichts),
            "csv" => Some(Self::Csv),
            "bbcode" | "bb" => Some(Self::Bbcode),
            "html" => Some(Self::Html),
            "emacs" => Some(Self::Emacs),
            "markdown" | "md" => Some(Self::Markdown),
            _ => None,
        }
    }

    pub fn syntax_markup(self) -> SyntaxMarkup {
        match self {
            Self::Nichts => SyntaxMarkup::empty(self),
            Self::Shell => SyntaxMarkup::empty(self),
            Self::Csv => SyntaxMarkup::empty(self),
            Self::Emacs => SyntaxMarkup {
                mode: self,
                begin_table: "".to_string(),
                end_table: "".to_string(),
                begin_cell: "|".to_string(),
                end_cell: "".to_string(),
                begin_row: "".to_string(),
                end_row: "|".to_string(),
            },
            Self::Markdown => SyntaxMarkup {
                mode: self,
                begin_table: "".to_string(),
                end_table: "".to_string(),
                begin_cell: "|".to_string(),
                end_cell: "".to_string(),
                begin_row: "".to_string(),
                end_row: "|".to_string(),
            },
            Self::Bbcode => SyntaxMarkup {
                mode: self,
                begin_table: "[table]".to_string(),
                end_table: "[/table]".to_string(),
                begin_cell: "[td]".to_string(),
                end_cell: "[/td]".to_string(),
                begin_row: "[tr]".to_string(),
                end_row: "[/tr]".to_string(),
            },
            Self::Html => SyntaxMarkup {
                mode: self,
                begin_table: r#"<table border=0 id="bigtable">"#.to_string(),
                end_table: "</table>\n".to_string(),
                begin_cell: "<td>\n".to_string(),
                end_cell: "\n</td>\n".to_string(),
                begin_row: "".to_string(),
                end_row: "</tr>\n".to_string(),
            },
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct SyntaxMarkup {
    pub mode: OutputMode,
    pub begin_table: String,
    pub end_table: String,
    pub begin_cell: String,
    pub end_cell: String,
    pub begin_row: String,
    pub end_row: String,
}

impl SyntaxMarkup {
    pub fn empty(mode: OutputMode) -> Self {
        Self {
            mode,
            begin_table: String::new(),
            end_table: String::new(),
            begin_cell: String::new(),
            end_cell: String::new(),
            begin_row: String::new(),
            end_row: String::new(),
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct OutputModeSpec {
    pub canonical_name: String,
    pub cli_value: String,
    pub syntax_class: String,
    pub force_one_table: bool,
    pub force_zero_width: bool,
    pub marks_html_or_bbcode: bool,
    pub aliases: Vec<String>,
}

impl OutputModeSpec {
    pub fn from_mode(mode: OutputMode) -> Self {
        let canonical_name = mode.canonical_name().to_string();
        Self {
            canonical_name: canonical_name.clone(),
            cli_value: canonical_name.clone(),
            syntax_class: mode.syntax_class_name().to_string(),
            force_one_table: mode.force_one_table(),
            force_zero_width: mode.force_zero_width(),
            marks_html_or_bbcode: mode.marks_html_or_bbcode(),
            aliases: vec![canonical_name],
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct OutputSyntaxSnapshot {
    pub class: String,
    pub modes: Vec<OutputModeSpec>,
    pub legacy_owner: String,
    pub architecture_owner: String,
}

#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub struct OutputSyntaxBundle;

impl OutputSyntaxBundle {
    pub fn modes(&self) -> Vec<OutputMode> {
        vec![
            OutputMode::Shell,
            OutputMode::Nichts,
            OutputMode::Csv,
            OutputMode::Bbcode,
            OutputMode::Html,
            OutputMode::Emacs,
            OutputMode::Markdown,
        ]
    }

    pub fn spec_for(&self, mode: OutputMode) -> OutputModeSpec {
        OutputModeSpec::from_mode(mode)
    }

    pub fn class_for(&self, mode: OutputMode) -> &'static str {
        mode.syntax_class_name()
    }

    pub fn snapshot(&self) -> OutputSyntaxSnapshot {
        OutputSyntaxSnapshot {
            class: "OutputSyntaxBundle".to_string(),
            modes: self.modes().into_iter().map(OutputModeSpec::from_mode).collect(),
            legacy_owner: "libs.lib4tables".to_string(),
            architecture_owner: "reta_architecture.output_syntax".to_string(),
        }
    }
}

pub fn bootstrap_output_syntax() -> OutputSyntaxBundle {
    OutputSyntaxBundle
}

pub fn colored_begin_col(mode: OutputMode, num: i64, rest: bool) -> String {
    match mode {
        OutputMode::Bbcode => colored_begin_col_bbcode(num, rest),
        OutputMode::Html => colored_begin_col_html(num, rest),
        _ => mode.syntax_markup().begin_row,
    }
}

fn colored_begin_col_bbcode(num: i64, rest: bool) -> String {
    let number_type = prime_creativity(num);
    if rest {
        return "[tr]".to_string();
    }
    match (number_type, num % 2 == 0, num) {
        (1, true, _) => r#"[tr="background-color:#66ff66;color:#000000;"]"#.to_string(),
        (1, false, _) => r#"[tr="background-color:#009900;color:#ffffff;"]"#.to_string(),
        (2, true, _) | (_, true, 1) => r#"[tr="background-color:#ffff66;color:#000099;"]"#.to_string(),
        (2, false, _) | (_, false, 1) => r#"[tr="background-color:#555500;color:#aaaaff;"]"#.to_string(),
        (3, true, _) => r#"[tr="background-color:#9999ff;color:#202000;"]"#.to_string(),
        (3, false, _) => r#"[tr="background-color:#000099;color:#ffff66;"]"#.to_string(),
        (_, _, 0) => r#"[tr="background-color:#ff2222;color:#002222;"]"#.to_string(),
        _ => "[tr]".to_string(),
    }
}

fn colored_begin_col_html(num: i64, rest: bool) -> String {
    let number_type = prime_creativity(num);
    if rest {
        return "<tr>\n".to_string();
    }
    match (number_type, num % 2 == 0, num) {
        (1, true, _) => r#"<tr style="background-color:#66ff66;color:#000000;">
"#.to_string(),
        (1, false, _) => r#"<tr style="background-color:#009900;color:#ffffff;">
"#.to_string(),
        (2, true, _) | (_, true, 1) => r#"<tr style="background-color:#ffff66;color:#000099;">
"#.to_string(),
        (2, false, _) | (_, false, 1) => r#"<tr style="background-color:#555500;color:#aaaaff;">
"#.to_string(),
        (3, true, _) => r#"<tr style="background-color:#9999ff;color:#202000;">
"#.to_string(),
        (3, false, _) => r#"<tr style="background-color:#000099;color:#ffff66;">
"#.to_string(),
        (_, _, 0) => r#"<tr style="background-color:#ff2222;color:#002222;">
"#.to_string(),
        _ => "<tr>\n".to_string(),
    }
}

pub fn generate_cell_begin(
    mode: OutputMode,
    spalte: i64,
    content: Option<i64>,
    zeile: Option<i64>,
    header_tags: &[String],
) -> String {
    match mode {
        OutputMode::Nichts => String::new(),
        OutputMode::Bbcode => {
            let adjusted = spalte + 2;
            let color = if adjusted == 0 {
                match content {
                    Some(value) if value % 2 == 0 => r#"="background-color:#000000;color:#ffffff""#,
                    Some(_) => r#"="background-color:#ffffff;color:#000000""#,
                    None => "=\"\"",
                }
            } else {
                "=\"\""
            };
            format!("[td{color}]")
        }
        OutputMode::Html => {
            let adjusted = spalte + 2;
            let mut attributes = String::new();
            if zeile == Some(0) {
                attributes.push_str(&format!(r#" class="z_{} r_{}""#, zeile.unwrap_or(0), adjusted));
            }
            if adjusted == 0 || adjusted == 1 {
                if let Some(value) = content {
                    if value % 2 == 0 {
                        attributes.push_str(r#" style="background-color:#000000;color:#ffffff;""#);
                    } else {
                        attributes.push_str(r#" style="background-color:#ffffff;color:#000000;""#);
                    }
                }
            } else if header_tags.iter().any(|tag| tag == "Symbole") {
                attributes.push_str(r#" class="tdSymbole" style="background-image: url();background-size: cover;background-repeat: no-repeat;background-position: right; ""#);
            }
            format!("<td{attributes}>\n")
        }
        _ => mode.syntax_markup().begin_cell,
    }
}

pub fn output_syntax_snapshot() -> OutputSyntaxSnapshot {
    bootstrap_output_syntax().snapshot()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn syntax_flags_match_python_classes() {
        assert!(OutputMode::Csv.force_one_table());
        assert!(OutputMode::Markdown.force_zero_width());
        assert!(OutputMode::Html.marks_html_or_bbcode());
        assert_eq!(OutputMode::from_name("md"), Some(OutputMode::Markdown));
    }

    #[test]
    fn colored_rows_follow_prime_creativity() {
        assert!(colored_begin_col(OutputMode::Html, 7, false).contains("009900"));
        assert!(colored_begin_col(OutputMode::Bbcode, 8, false).contains("9999ff"));
        assert_eq!(colored_begin_col(OutputMode::Csv, 8, false), "");
    }

    #[test]
    fn cell_begin_uses_mode_syntax() {
        assert_eq!(generate_cell_begin(OutputMode::Markdown, 1, None, None, &[]), "|");
        assert!(generate_cell_begin(OutputMode::Html, -2, Some(2), Some(1), &[]).contains("td"));
    }
}
